enterprise/server/cmd/action_runner/main.go

```diff
@@ -0,0 +1,662 @@
+	if err != nil {
+		return err
+	}
+	if err := runCommand(ctx, "git", []string{"remote", "add", "origin", authURL}, emptyEnv, os.Stderr); err != nil {
+		return err
+	}
+	if err := runCommand(ctx, "git", []string{"fetch", "origin", *commitSHA}, emptyEnv, os.Stderr); err != nil {
+		return err
+	}
+	if err := runCommand(ctx, "git", []string{"checkout", *commitSHA}, emptyEnv, os.Stderr); err != nil {
+		return err
+	}
+	return nil
+}
+
+func authRepoURL() (string, error) {
+	user := os.Getenv(repoUserEnvVarName)
+	token := os.Getenv(repoTokenEnvVarName)
+	if user == "" && token == "" {
+		return *repoURL, nil
+	}
+	u, err := url.Parse(*repoURL)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse repo URL %q: %s", *repoURL, err)
+	}
+	u.User = url.UserPassword(user, token)
+	return u.String(), nil
+}
+
+func invocationURL(invocationID string) string {
+	urlPrefix := *besResultsURL
+	if !strings.HasSuffix(urlPrefix, "/") {
+		urlPrefix = urlPrefix + "/"
+	}
+	return urlPrefix + invocationID
+}
+
+func readConfig() (*workflowconf.BuildBuddyConfig, error) {
+	f, err := os.Open(actionsConfigPath)
+	if err != nil {
+		return nil, fmt.Errorf("open %q: %s", actionsConfigPath, err)
+	}
+	c, err := workflowconf.NewConfig(f)
+	if err != nil {
+		return nil, fmt.Errorf("read %q: %s", actionsConfigPath, err)
+	}
+	return c, nil
+}
+
+func matchesAnyTrigger(action *workflowconf.Action, event, branch string) bool {
+	if action.Triggers == nil {
+		return false
+	}
+	if pushCfg := action.Triggers.Push; pushCfg != nil && event == pushEventName {
+		return matchesAnyBranch(pushCfg.Branches, branch)
+	}
+
+	if prCfg := action.Triggers.PullRequest; prCfg != nil && event == pullRequestEventName {
+		return matchesAnyBranch(prCfg.Branches, branch)
+	}
+	return false
+}
+
+func matchesAnyBranch(branches []string, branch string) bool {
+	if branches == nil {
+		return false
+	}
+	for _, b := range branches {
+		if b == branch {
+			return true
+		}
+	}
+	return false
+}
+
+func runCommand(ctx context.Context, executable string, args []string, env map[string]string, outputSink io.Writer) error {
+	cmd := exec.CommandContext(ctx, executable, args...)
+	cmd.Stdout = outputSink
+	cmd.Stderr = outputSink
+	for k, v := range env {
+		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", k, v))
+	}
+	return cmd.Run()
+}
+
+func getExitCode(err error) int {
+	if err == nil {
+		return 0
+	}
+	if exitErr, ok := err.(*exec.ExitError); ok {
+		return exitErr.ExitCode()
+	}
+	return noExitCode
+}
+
+func validateFlags() error {
+	errs := []string{}
+	for name, flag := range map[string]*string{
+		"repo_url":        repoURL,
+		"commit_sha":      commitSHA,
+		"trigger_branch":  triggerBranch,
+		"trigger_event":   triggerEvent,
+		"bes_backend":     besBackend,
+		"bes_results_url": besResultsURL,
+	} {
+		if *flag == "" {
+			errs = append(errs, fmt.Sprintf("--%s unset", name))
+		}
+	}
+	if len(errs) == 0 {
+		return nil
+	}
+	return fmt.Errorf(strings.Join(errs, "; "))
+}
```

Comment:
I would probably just remove this -- most of the stuff (like conn.Dial) will fail sanely if you don't set a flag.

Generally, it works better to validate args before you use them. That allows:
 - other code to import yours and still work if flags are unset (doesn't apply here I know)
 - better validation of the flag, it could be conditional, or you could log an error message with some other computed data.

Reply:
Good point, and thanks for the explanation. This made more sense when I was developing and wanted to fail early with a blatantly obvious error message if I forgot to set a flag, but now it's not so useful -- removed